### PR TITLE
feat(p2p): moving gossiped blocks reception from validator to p2p client

### LINKED ISSUE #802: decouple p2p validation from block reception and application
applying blocks to the store should not block the p2p validation

### PR DESCRIPTION
**Pull request by srene**
# PR Standards

This PR decouples p2p gossip blocks validation from blocks reception. By doing that we avoid any interference of block store application in block propagation.

## Opening a pull request should be able to meet the following requirements

---

Close #802 

<-- Briefly describe the content of this pull request -->

For Author:

- [ ]  Targeted PR against correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to Github issue with discussion and accepted design
- [ ]  Targets only one github issue
- [ ]  Wrote unit and integration tests
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)

---

For Reviewer:

- [ ]  confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  confirmed all author checklist items have been addressed

---

After reviewer approval:

- [ ]  In case targets main branch, PR should be squashed and merged.
- [ ]  In case PR targets a release branch, PR should be rebased.

### CODE REVIEW

File: p2p/validator.go [hunk 1]
```diff
@@ -82,15 +78,15 @@ func (v *Validator) BlockValidator() GossipValidator {
...
 			return false
 		}
 
-		err := v.localPubsubServer.PublishWithEvents(context.Background(), gossipedBlock, map[string][]string{EventTypeKey: {EventNewGossipedBlock}})
+		/*err := v.localPubsubServer.PublishWithEvents(context.Background(), gossipedBlock, map[string][]string{EventTypeKey: {EventNewGossipedBlock}})
 		if err != nil {
 			v.logger.Error("publishing event", "err", err)
 			return false
-		}
+		}*/
```
Comment (danwt) [inline p2p/validator.go:89]: @srene supposed to be commented out?

Comment (srene) [reply p2p/validator.go]: yes, but i forgot to remove it. done

File: p2p/validator.go [hunk 2]
```diff
@@ -82,15 +78,10 @@ func (v *Validator) BlockValidator() GossipValidator {
 			return false
 		}
 		if err := gossipedBlock.Validate(v.slClient.GetProposer()); err != nil {
-			v.logger.Error("Invalid gossiped block", "error", err)
+			v.logger.Error("Invalid gossiped block", "height", gossipedBlock.Block.Header.Height, "error", err)
```
Comment (danwt) [inline p2p/validator.go:81]:
```suggestion
			v.logger.Error("Invalid gossiped block.", "height", gossipedBlock.Block.Header.Height, "error", err)
```

File: p2p/gossip.go [hunk 3]
```diff
@@ -22,6 +22,8 @@ type GossipMessage struct {
 // GossiperOption sets optional parameters of Gossiper.
 type GossiperOption func(*Gossiper) error
 
+type NewMessage func(msg *GossipMessage)
```
Comment (danwt) [inline p2p/gossip.go:25]: can we rename this `GossipMessageHandler`

File: p2p/gossip.go [hunk 4]
```diff
@@ -33,17 +35,17 @@ func WithValidator(validator GossipValidator) GossiperOption {
...
 
-	ps    *pubsub.PubSub
-	topic *pubsub.Topic
-	sub   *pubsub.Subscription
-
-	logger types.Logger
+	ps          *pubsub.PubSub
+	topic       *pubsub.Topic
+	sub         *pubsub.Subscription
+	msgReceiver NewMessage
```
Comment (danwt) [inline p2p/gossip.go:41]: can we rename to `msgHandler`

File: p2p/gossip.go [hunk 5]
```diff
@@ -33,17 +35,17 @@ func WithValidator(validator GossipValidator) GossiperOption {
...
+	logger      types.Logger
 }
 
 // NewGossiper creates new, ready to use instance of Gossiper.
 //
 // Returned Gossiper object can be used for sending (Publishing) and receiving messages in topic identified by topicStr.
-func NewGossiper(host host.Host, ps *pubsub.PubSub, topicStr string, logger types.Logger, options ...GossiperOption) (*Gossiper, error) {
+func NewGossiper(host host.Host, ps *pubsub.PubSub, topicStr string, msgReceiver NewMessage, logger types.Logger, options ...GossiperOption) (*Gossiper, error) {
```
Comment (danwt) [inline p2p/gossip.go:48]: rename msgHandler

**Review by danwt — CHANGES_REQUESTED** [review]
So I understand it's a simple change to move the local gossip part out of the validator, and into the message handler?
Looks good but I requested some renames

File: p2p/gossip.go (hunk 5 above)
Comment (srene) [reply p2p/gossip.go]: done

File: p2p/gossip.go (hunk 4 above)
Comment (srene) [reply p2p/gossip.go]: done

File: p2p/gossip.go (hunk 3 above)
Comment (srene) [reply p2p/gossip.go]: done

**Review by danwt — APPROVED** [review]
great work!

**Review by mtsitrin — APPROVED** [review]

**Pull request merged**